### PR TITLE
[Doc] Single-machine Multi-GPU node classification tutorial

### PR DESCRIPTION
**Pull request by BarclayII**
Also included a bunch of changes in the DataLoader to make DistributedSampler work with GPUs.

@nv-dlasalle maybe you can check if my changes to the DataLoader make sense?

### CODE REVIEW

File: tutorials/multi/2_node_classification.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+# which *forks* from the main process and allows sharing the same graph
+# object to trainer processes via *copy-on-write*, as long as the graph
+# object is not mutated. This can greatly reduce the memory consumption.
+# 
+# To keep the graph object from mutating, you need to call
+# ``create_formats_`` so that no new sparse matrix representations of the
+# graph will be created on the fly.
+# 
+
+graph.create_formats_()
```
Comment (hetong007) [inline tutorials/multi/2_node_classification.py:223]: Do we have a better name? This `create_formats_` looks internal.

File: tutorials/multi/2_node_classification.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+# 
+# .. note::
+# 
+#    You will need to use ``dgl.multiprocessing`` instead of the Python
+#    ``multiprocessing`` package. ``dgl.multiprocessing`` is identical to
+#    Python’s built-in ``multiprocessing`` except that it handles the
+#    subtleties between forking and multithreading in Python.
+# 
+
+# Say you have four GPUs.
```
Comment (hetong007) [inline tutorials/multi/2_node_classification.py:237]: Can we simply define a `num_gpus`?

File: tutorials/multi/2_node_classification.py [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+    sampler = dgl.dataloading.MultiLayerNeighborSampler([4, 4])
+    train_dataloader = dgl.dataloading.NodeDataLoader(
+        # The following arguments are specific to NodeDataLoader.
+        graph,              # The graph
+        train_nids,         # The node IDs to iterate over in minibatches
+        sampler,            # The neighbor sampler
+        device=device,      # Put the sampled MFGs on CPU or GPU
+        use_ddp=True,       # Make it work with distributed data parallel
+        # The following arguments are inherited from PyTorch DataLoader.
+        batch_size=1024,    # Batch size
```
Comment (hetong007) [inline tutorials/multi/2_node_classification.py:125]: Is it per-device batch_size? Maybe we need to address that the total effective batch size is this `batch_size * num_devices`.

File: tutorials/multi/1_graph_classification.py [hunk 4]
```diff
@@ -221,10 +226,17 @@ def main(rank, world_size, dataset, seed=0):
...
+# 
+#    You will need to use ``dgl.multiprocessing`` instead of the Python
+#    ``multiprocessing`` package. ``dgl.multiprocessing`` is identical to
+#    Python’s built-in ``multiprocessing`` except that it handles the
+#    subtleties between forking and multithreading in Python.
 #
 
 if __name__ == '__main__':
-    import torch.multiprocessing as mp
+    import dgl.multiprocessing as mp
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply tutorials/multi/1_graph_classification.py:239]: With MX/TF it fallbacks to Python's builtin multiprocessing.

File: tutorials/multi/2_node_classification.py (hunk 1 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:223]: The suffix underscore represents an in-place operation like PyTorch.  It's been there long ago and is used in a couple places already.  I would rather keep it like this.

File: tutorials/multi/2_node_classification.py (hunk 3 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py]: Done.

File: tutorials/multi/2_node_classification.py (hunk 2 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:238]: Done.

File: tutorials/multi/2_node_classification.py (hunk 1 above)
Comment (hetong007) [reply tutorials/multi/2_node_classification.py:224]: I'm not familiar with this API, thus its function here is less intuitive to me: there's a gap between "create new sparse mat rep" and "keep it from mutating". Why is it immutable (is it?) when all sparse formats are created?

File: tutorials/multi/2_node_classification.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                opt.step()
+
+                accuracy = sklearn.metrics.accuracy_score(labels.cpu().numpy(), predictions.argmax(1).detach().cpu().numpy())
+
+                tq.set_postfix({'loss': '%.03f' % loss.item(), 'acc': '%.03f' % accuracy}, refresh=False)
+
+        model.eval()
+
+        # Evaluate on only the first GPU.
+        if proc_id == 0:
```
Comment (nv-dlasalle) [inline tutorials/multi/2_node_classification.py:179]: @BarclayII I've noticed in the other DGL examples, validation/testing is often done on just one GPU. What's the reason for this? In the multi-GPU cases, it's generally slower than the training due to this.

File: tutorials/multi/2_node_classification.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+# 
+
+def run(proc_id, devices):
+    # Initialize distributed training context.
+    dev_id = devices[proc_id]
+    dist_init_method = 'tcp://{master_ip}:{master_port}'.format(master_ip='127.0.0.1', master_port='12345')
+    # To use Multi-GPU, change 'gloo' to 'nccl'.
+    torch.distributed.init_process_group(
+        backend='gloo', init_method=dist_init_method, world_size=len(devices), rank=proc_id)
+    device = torch.device('cpu')
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:108]: This is really a multi-GPU tutorial but our CI will test-run the tutorials as well, and the CI that runs the tutorials only has CPU.  So it's not quite possible to default to GPU.

File: tutorials/multi/2_node_classification.py (hunk 5 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:179]: I'm not sure what is an elegant way to perform multi-GPU validation and testing because I will need to send the results to a single process where the final validation/test accuracy is computed.  I remember RGCN has multi-GPU validation but the code is quite tedious (it has a queue and something).

File: tutorials/multi/2_node_classification.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        h = self.conv2(mfgs[1], (h, h_dst))
+        return h
+
+
+######################################################################
+# Defining Training Procedure
+# ---------------------------
+# 
+# The training procedure will be slightly different from what you saw
+# previously, in the sense that you will need to \* Initialize a
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py]: Fixed.

File: python/dgl/dataloading/pytorch/__init__.py [hunk 8]
```diff
@@ -62,16 +93,52 @@ def __iter__(self):
...
+    def _iter_ddp(self):
+        # The following code (and the idea of cross-process shuffling with the same seed)
+        # comes from PyTorch.  See torch/utils/data/distributed.py for details.
+        if self.shuffle:
+            # deterministically shuffle based on epoch and seed
+            g = th.Generator()
+            g.manual_seed(self.seed + self.epoch)
+            indices = th.randperm(len(self.dataset), generator=g).tolist()  # type: ignore
+        else:
+            indices = list(range(len(self.dataset)))  # type: ignore
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py]: Fixed.

I copied this from PyTorch.  Not sure why they use a list either.

File: tutorials/multi/2_node_classification.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+                    predictions.append(model(mfgs, inputs).argmax(1).cpu().numpy())
+                predictions = np.concatenate(predictions)
+                labels = np.concatenate(labels)
+                accuracy = sklearn.metrics.accuracy_score(labels, predictions)
+                print('Epoch {} Validation Accuracy {}'.format(epoch, accuracy))
+                if best_accuracy < accuracy:
+                    best_accuracy = accuracy
+                    torch.save(model.state_dict(), best_model_path)
+
+        # Note that this tutorial do not train the whole model to the end.
```
[the remark this answers is not in the document]
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:196]: I guess it's OK as our previous large graph training tutorials are doing the same thing.

File: tutorials/multi/2_node_classification.py (hunk 1 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:232]: Elaborated a bit more.  See if this makes sense?

Comment (hetong007) [reply tutorials/multi/2_node_classification.py:232]: Looks much better, thanks! Separately, the topic of formats could be a vignette in user guide, if we don't yet have one.

**Review by hetong007 — APPROVED** [review]

File: python/dgl/dataloading/pytorch/__init__.py [hunk 10]
```diff
@@ -42,16 +45,44 @@ class _ScalarDataBatcher(th.utils.data.IterableDataset):
...
+        if use_ddp:
+            self.rank = dist.get_rank()
+            self.num_replicas = dist.get_world_size()
+            self.seed = ddp_seed
+            self.epoch = 0
+            # The following code (and the idea of cross-process shuffling with the same seed)
+            # comes from PyTorch.  See torch/utils/data/distributed.py for details.
+
+            # If the dataset length is evenly divisible by # of replicas, then there
+            # is no need to drop any data, since the dataset will be split equally.
```
Comment (mufeili) [inline python/dgl/dataloading/pytorch/__init__.py:64]: "data" -> "sample"

Comment (mufeili) [inline python/dgl/dataloading/pytorch/__init__.py:64]: "equally" -> "evenly"?

File: tutorials/multi/2_node_classification.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
+"""
+Single Machine Multi-GPU Minibatch Node Classification
+======================================================
+
+In this tutorial, you will learn how to use multiple GPUs in training a
+graph neural network (GNN) for graph classification.
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:6]: "graph classification" -> "node classification"

File: tutorials/multi/2_node_classification.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+This tutorial assumes that you have read the :doc:`Training GNN with Neighbor
+Sampling for Node Classification <../large/L1_large_node_classification>`
+tutorial. It also assumes that you know the basics of training general
+models with multi-GPU with ``DistributedDataParallel``.
+
+.. note::
+
+   See :doc:`the multi-GPU graph classification
+   tutorial <1_graph_classification>`
+   for an overview of using ``DistributedDataParallel`` with DGL.
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:19]: Can we tell readers which specific paragraphs/section they will need to read?

File: tutorials/multi/2_node_classification.py [hunk 13]
```diff
@@ -0,0 +1,257 @@
...
+
+"""
+
+
+######################################################################
+# Loading Dataset
+# ---------------
+# 
+# OGB already prepared the data as DGL graph. The following code is
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:28]: "DGL graph" -> "a DGLGraph" or "DGLGraphs".

File: tutorials/multi/2_node_classification.py [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+graph.ndata['label'] = node_labels[:, 0]
+
+node_features = graph.ndata['feat']
+num_features = node_features.shape[1]
+num_classes = (node_labels.max() + 1).item()
+
+idx_split = dataset.get_idx_split()
+train_nids = idx_split['train']
+valid_nids = idx_split['valid']
+test_nids = idx_split['test']
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:58]: Not used

File: tutorials/multi/2_node_classification.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mufeili) [reply tutorials/multi/2_node_classification.py:109]: Alternatively, you can add if __name__ == '__main__' so that CI will not test this tutorial.

File: python/dgl/dataloading/pytorch/__init__.py (hunk 10 above)
Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py]: I copied the comments from PyTorch haha.

Comment (BarclayII) [reply python/dgl/dataloading/pytorch/__init__.py]: Ditto.

File: tutorials/multi/2_node_classification.py (hunk 12 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py:21]: Done.

File: tutorials/multi/2_node_classification.py (hunk 11 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py]: Done.

File: tutorials/multi/2_node_classification.py (hunk 13 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py]: Done.

File: tutorials/multi/2_node_classification.py (hunk 14 above)
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py]: I kept it regardless but added a comment saying that it is not used.

File: tutorials/multi/2_node_classification.py [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+    )
+    valid_dataloader = dgl.dataloading.NodeDataLoader(
+        graph, valid_nids, sampler,
+        batch_size=1024,
+        shuffle=False,
+        drop_last=False,
+        num_workers=0,
+        device=device,
+        use_ddp=False
+    )
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:142]: It's better to use a consistent order of the arguments for both `train_dataloader` and `valid_dataloader`.

File: tutorials/multi/2_node_classification.py [hunk 16]
```diff
@@ -0,0 +1,259 @@
...
+                    predictions.append(model(mfgs, inputs).argmax(1).cpu().numpy())
+                predictions = np.concatenate(predictions)
+                labels = np.concatenate(labels)
+                accuracy = sklearn.metrics.accuracy_score(labels, predictions)
+                print('Epoch {} Validation Accuracy {}'.format(epoch, accuracy))
+                if best_accuracy < accuracy:
+                    best_accuracy = accuracy
+                    torch.save(model.state_dict(), best_model_path)
+
+        # Note that this tutorial do not train the whole model to the end.
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:198]: "do" -> "does"

File: tutorials/multi/2_node_classification.py [hunk 17]
```diff
@@ -0,0 +1,259 @@
...
+# multiple processes. This however is undesirable for training node
+# classification or link prediction models on a single large graph,
+# especially on Linux. The reason is that a single large graph itself may
+# take a lot of memory, and ``mp.spawn`` will duplicate all objects in the
+# program, including the large graph. Consequently, the large graph will
+# be duplicated as many times as the number of GPUs.
+# 
+# To alleviate the problem we recommend using ``multiprocessing.Process``,
+# which *forks* from the main process and allows sharing the same graph
+# object to trainer processes via *copy-on-write*. This can greatly reduce
```
Comment (mufeili) [inline tutorials/multi/2_node_classification.py:219]: "to" -> "with"?

File: tutorials/multi/2_node_classification.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (nv-dlasalle) [reply tutorials/multi/2_node_classification.py:111]: What if we had check for the number of GPUs, and only run on the CPU if none are detected? Something like?
```
if torch.cuda.device_count() < 1:
  print("WARNING: No CUDA capable devices found, running on the CPU."
  backend='gloo'
  device = torch.device('cpu')
```

[the remark this answers is not in the document]
Comment (BarclayII) [reply tutorials/multi/2_node_classification.py]: K.
